### DISCUSSION TITLE
Quarkus/JSON-B/Yasson and Blaze's Keyset Pagination

### DISCUSSION THREAD

**Discussion by Akaame** [Blazebit/blaze-persistence#1470] [Q&A] [answered]
Hello,

I am kind of stuck with how one would go about de/serialization of `KeysetPage` and `Keyset` itself. Documentation suggests creating custom implementation so that custom JSON handling can be made since `Serializable` itself has no type info and canniot be deserialized. To that end, I was planning to create `CustomKeyset` which also includes the type in the tuple so that we can actually convert it back its original type. However, then you need to handle every possible column type (at least the ones you are using) in your deserializer. Is this the recommended way of working with Keyset Pagination for REST or is there a facility that I need to use and am not aware of.?

The Quarkus JAX-RS Jackson example does not have Keyset Pagination and there is no JSON-B example. Would it possible for you to create these examples showcasing the recommended way of working with the library?

Cheers

**Comment by beikov** [comment, marked as answer]
Hi,

the serialization part was deliberately not handled out of the box because everyone probably has a different opinion about how this should be done. For GraphQL we use Java serialization with a serialization filter, so that only the basic types of the entity domain can be deserialized. You could use something similar. Also see https://github.com/Blazebit/blaze-persistence/blob/main/integration/graphql/src/main/java/com/blazebit/persistence/integration/graphql/GraphQLRelayPageInfo.java#L60 and https://github.com/Blazebit/blaze-persistence/blob/main/integration/graphql/src/main/java/com/blazebit/persistence/integration/graphql/GraphQLCursorObjectInputStream.java

I honestly don't know how to best model this in Quarkus native though. Is serialization supported now in native images? You can alternatively serialize the keyset into JSON though and include the type in the payload, as you suggested, though you will probably need something similar to the deserialization filter and I am not sure all JSON libraries provide an easy way to do this. Also see: https://github.com/Blazebit/blaze-persistence/blob/main/integration/graphql/src/main/java/com/blazebit/persistence/integration/graphql/GraphQLEntityViewSupportFactory.java#L476

**Reply by Akaame** [reply]
Servus,

> I honestly don't know how to best model this in Quarkus native though. Is serialization supported now in native images? 

We are running Quarkus over JVM so I cannot answer that. 

So what I ended up doing is pretty similar to your GraphQL + java.io.Serialiation solution.

- Created barebones impls for `KeysetPage` and co.
- Instead of using a `Serializable[]` created a class which is basically a (type, value) pair.
- Created a custom JSON-B deserializer (`JsonbDeserializer<T>`) for the said class allowing only UUID, String, Integer and Long types (which should cover most of the simple key scenarios) would be deserialized.

Can't say I am really proud of my impl and I can see why people would have different choices here. But, I think, a Quarkus example REST api with keyset pagination would be greatly appreciated.

Cheers.

**Reply by beikov** [reply]
Great to hear that you figured it out. If you could share your work somehow, I can try to integrate that into the example app.
Maybe we can add that into the various JSON integrations and leverage that from within the other integrations i.e. Spring/Quarkus etc.
